packages/next-app/src/hooks/useTransaction.ts

```diff
@@ -294,118 +293,126 @@ export function useInternalBalancesTransaction({
+          NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.APPROVED]
         );
       }
     } catch (error) {
       setNotification(
-        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.PINNING_ERROR]
       );
       setTransactionStatus(TransactionStatus.AUTHORIZING);
     }
   }
 
-  function handleTransaction({
-    data,
-    decimals,
-  }: {
-    data: FieldValues;
-    decimals: number;
-  }) {
-    setSubmitData([
-      {
-        tokenAddress: data.tokenAddress,
-        tokenAmount: data.tokenAmount,
-        tokenDecimals: decimals,
-        receiverAddress: data.receiverAddress,
-      },
-    ]);
-    refetch();
-  }
-
-  useEffect(() => {
-    if (submitData.length === 0) return;
+  async function handleTransaction({ data }: { data: SubmitData[] }) {
     setTransactionUrl(undefined);
-    setNotification(
-      NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.AUTHORIZING]
-    );
     if (
       operationKind === UserBalanceOpKind.DEPOSIT_INTERNAL &&
       transactionStatus === TransactionStatus.AUTHORIZING
-    ) {
-      approveToken();
-    } else {
-      if (!write) return;
-      try {
-        setNotification(
-          NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.AUTHORIZING]
-        );
-        write();
-        setTransactionStatus(TransactionStatus.SUBMITTING);
-      } catch (error) {
-        setNotification(
-          NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
-        );
-        setTransactionStatus(TransactionStatus.AUTHORIZING);
-      }
-    }
-  }, [submitData]);
-
-  useEffect(() => {
-    if (!error) return;
-    if (
-      operationKind === UserBalanceOpKind.DEPOSIT_INTERNAL &&
-      transactionStatus === TransactionStatus.SUBMITTING
     ) {
       setNotification(
-        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.AUTHORIZING]
       );
-      setTransactionStatus(TransactionStatus.CONFIRMING);
+      approveToken({ data });
     } else {
       setNotification(
-        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WRITE_ERROR]
+        NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.CONFIRMING]
       );
-      setTransactionStatus(TransactionStatus.AUTHORIZING);
+      try {
+        setTransactionStatus(TransactionStatus.SUBMITTING);
+        const userBalancesOp = data.map((transactionItem: SubmitData) => {
+          return {
+            kind: operationKind as number,
+            sender: userAddress,
+            recipient: transactionItem.receiverAddress as Address,
+            asset: transactionItem.tokenAddress,
+            amount: parseUnits(
+              transactionItem.tokenAmount,
+              transactionItem.tokenDecimals
+            ),
+          };
+        });
+        const config = await prepareWriteContract({
```

Comment:
I know this was already here but want to reiterate my suggestion to keep domain-related stuff somewhere else. This hook is so complex already -- adding domain-specific code just makes it even harder to reason about it.

Reply:
in this case, the domain-related would be calling the contract?